packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -441,38 +395,110 @@ export class ScheduleManager {
 
         // If no message has caused the pause flag to be set, or the next message up is not the one we need to pause at
         // then we simply continue processing
-        if (!this.pauseSequenceNumber || sequenceNumber + 1 < this.pauseSequenceNumber) {
+        if (this.pauseSequenceNumber === undefined || sequenceNumber + 1 < this.pauseSequenceNumber) {
             this.setPaused(false);
         } else {
             // Otherwise the next message requires us to pause
             this.setPaused(true);
         }
     }
 
+    /**
+     * Called for each incoming op (i.e. inbound "push" notification)
+     */
     private trackPending(message: ISequencedDocumentMessage) {
-        const metadata = message.metadata as IRuntimeMessageMetadata | undefined;
+        const metadata = message.metadata as IRuntimeMessageMetadata;
 
         // Protocol messages are never part of a runtime batch of messages
         if (!isRuntimeMessage(message)) {
             this.pauseSequenceNumber = undefined;
-            this.pauseClientId = undefined;
+            this.currentBatchClientId = undefined;
             return;
         }
 
         const batchMetadata = metadata ? metadata.batch : undefined;
 
         // If the client ID changes then we can move the pause point. If it stayed the same then we need to check.
-        if (this.pauseClientId === message.clientId) {
-            if (batchMetadata !== undefined) {
-                // If batchMetadata is not undefined then if it's true we've begun a new batch - if false we've ended
-                // the previous one
-                this.pauseSequenceNumber = batchMetadata ? message.sequenceNumber : undefined;
-                this.pauseClientId = batchMetadata ? this.pauseClientId : undefined;
-            }
-        } else {
-            // We check the batch flag for the new clientID - if true we pause otherwise we reset the tracking data
+        // If batchMetadata is not undefined then if it's true we've begun a new batch - if false we've ended
+        // the previous one
+        // We check the batch flag for the new clientID - if true we pause otherwise we reset the tracking data
+        if (this.currentBatchClientId !== message.clientId || batchMetadata !== undefined) {
             this.pauseSequenceNumber = batchMetadata ? message.sequenceNumber : undefined;
-            this.pauseClientId = batchMetadata ? message.clientId : undefined;
+            this.currentBatchClientId = batchMetadata ? message.clientId : undefined;
+        }
+    }
+}
+
+/**
+ * This class has the following responsibilities:
+ * 1. It tracks batches as we process ops and raises "batchBegin" and "batchEnd" events.
+ *    As part of it, it validates batch correctness (i.e. no system ops in the middle of batch)
+ * 2. It creates instance of ScheduleManagerCore that ensures we process ops in a batch when we have full batch in.
+ *    It notifies it about
+ */
+export class ScheduleManager {
```

Comment:
Can we move ScheduleManager and ScheduleManagerCore into a separate file?

Reply:
I'd prefer to do it as separate PR, to allow at least some inspection of code changes (as I try to keep them mechanical for most part, but they are not truly mechanical and I made a mistake in first commit :))